### ISSUE TITLE
Discussion about formatting "rules" (i.e. the one with parenthesis)

### ISSUE THREAD

**Issue by ericclemmons** [prettier/prettier#2180]
_Note: this is primarily for others who've used prettier since its inception to understand, as formatting styles have changed, the underlying rationale behind formatting decisions._

tldr; **How would you describe the "rules" that prettier operates against when deciding to keep or remove structural syntax?**.

---

### Background

_(Related to #1733 and now #2065)_

This tweet got some attention from some notable folks:
> https://twitter.com/ericclemmons/status/876179258888081408
> "I like prettier, but I’d prefer that parenthesis didn’t change instances like these." – @ericclemmons
> <img width="367" alt="screen shot 2017-06-17 at 3 41 02 pm" src="https://user-images.githubusercontent.com/15182/27256584-67fa2922-537f-11e7-9bb0-696f947b8655.png">
> <img width="490" alt="screen shot 2017-06-17 at 3 41 18 pm" src="https://user-images.githubusercontent.com/15182/27256583-67f9db7a-537f-11e7-86d0-29d5eebec53f.png">

### Rationale

My take on Prettier's "rationale" (rules) for formatting is to **balance legibility with consistency**.

Based on the changes since `1.3.1` (IIRC), it's more like **balance legibility with ~consistency~ _simplicity_**.

Early on, my sense was that ArrowExpressions generally matched the following:

- Single-line expressions with implicit return:

  ```js
  () => 42
  ```
- Multi-line expressions with implicit return:

  ```js
  (props) => (
    <div>
      Hello there, {props.children}!
    </div>
  );`
  ```

- Multi-line expressions with explicit return:

  ```js
  (props) => {
    const { children } = props;

    return (
      <div>
        Hello there, {children}!
      </div>
    );
  }`
  ```

As complexity grew (going from `()` to `(props)` to `({ children })`, etc.), formatting changes were minimal.

It seems now that there's a bit more of a _jello effect_, where it wiggles in more places than where you poked it ;)

**Comment by lydell** [comment]
> It seems now that there's a bit more of a jello effect, and so I had the following question:

It seems like there's nothing more after that colon. Did part of your message get cut off?

**Comment by ericclemmons** [comment]
(@lydell Whoops!  Yes, I fixed that last sentence.  Re-formatted it heavily.)

**Comment by Haroenv** [comment]
I’d prefer parenthesis for all multi-line arrow functions as well. Before prettier I was not aware that it was possible to do that, and it looks quite scary to me. The parenthesis around jsx also made it easier to move parts, since it’s completely separated from other parts of the code.

**Comment by btnwtn** [comment]
I want enforced parens for single parameter arrow functions so badly 👍

**Comment by kumarharsh** [comment]
+1 - it's one of those coffee-script things which I loved at the start, but after prolonged usage I discovered that it harmed the project much more than any benefit it had. There should be a way to preserve parenthesis.

**Comment by Anahkiasen** [comment]
Would love a way to enforce parenthesis usage for multiline arrow functions as well, it's the one thing I'm currently disagreeing on with Prettier's output

**Comment by vjeux** [comment]
I wonder how it would feel to have arrow functions that fit in a single line to have no parenthesis and if they don't fit, then use it.

**Comment by Anahkiasen** [comment]
That's how I did it before Prettier so it'd be more than fine for me, would be perfect actually :p

**Comment by k15a** [comment]
@vjeux I feel like this would be even more inconsistent than the current behavior.

**Comment by azz** [comment]
I guess this runs on the assumption that a function with a multiline body (whether a block out not?) will have a higher likelihood of being more complex and needing more arguments in the future?

It does feel super workaround-y and somewhat difficult to explain. It also may mean if you add a few more characters to an existing arrow function body it may cause parens to be added, which is odd UX.

Implementing the Airbnb suggestion might be the least worst option to appeal to the largest number of people.

**Comment by azz** [comment]
I guess this runs on the assumption that a function with a multiline body (whether a block out not?) will have a higher likelihood of being more complex and needing more arguments in the future?

It does feel super workaround-y and somewhat difficult to explain. It also may mean if you add a few more characters to an existing arrow function body it may cause parens to be added, which is odd UX.

Implementing the Airbnb suggestion might be the least worst option to appeal to the largest number of people.

**Comment by azz** [comment]
Disregard previous comment, thought it was regarding argument lists, not function bodies.

**Comment by TomiS** [comment]
One argument for adding parenthesis for single argument arrow functions would be better compatibility with flow. Let me explain.

Let's say you have a function that is automatically cleaned from having parenthesis:
```
const func = props => true
```
Now let's say you want to add flow typings for props and you write this
```
const funct = props: Object => true
```
This is an error in prettier because it cannot add parenthesis automatically. You need to either realise this upfront or go back and add parenthesis afterwards like this
```
const funct = (props: Object) => true
```
This is not an end of the world but it's a bit annoying.

Also, I cannot avoid expressing my opinion that I would love the consistency of having parenthesis in all arrow functions instead of 'randomly' skipping the one with single argument.

**Comment by Anahkiasen** [comment]
I think people are talking about two different things, @vjeux was talking about 

```js
const foobar = (bar) => 
    <bar>
         {bar}
    </bar>
```

versus

```js
const foobar = (bar) => (
    <bar>
         {bar}
    </bar>
);
```

no?

**Comment by azz** [comment]
Yep, we're probably getting confused due to this image in OP

![](https://user-images.githubusercontent.com/15182/27256583-67f9db7a-537f-11e7-86d0-29d5eebec53f.png)

Parameter list discussion is here: #812

**Comment by ssalka** [comment]
Has there been any movement on this issue? I know parentheses for multiline return values are preserved for JSX, but I also want to be able to preserve parentheses for any data type. for example, I do the same thing w/ lodash chains sometimes:

```
const getFormattedNames = arr => (
  _(arr)
    .flatMap('data.name')
    .compact()
    .map(_.startCase)
    .value()
);
```

From reading through this issue (and #2065), it seems that Prettier would still output this:

```
const getFormattedNames = arr =>
  _(arr)
    .flatMap('data.name')
    .compact()
    .map(_.startCase)
    .value();
```

This is the _only thing_ that is keeping me from using Prettier in all my projects. Would love to see support for this added!

**Comment by ericclemmons** [comment]
Closing due to #2208 & being shipped in `v1.6`.

**Comment by taion** [comment]
I do not believe this issue should have been closed. As far as I'm aware Prettier still has no option to wrap multiline arrow function bodies (or other bodies) in parentheses outside of JSX. See [Playground](https://prettier.io/playground/#N4Igxg9gdgLgprEAuc0DOMAEAzCFMC8mwAOlJpgLYCuMAhvAIJQAmA4nDAAp0CeANhDoskmABQBLKAAdaAGkyRYcAB4wAlIQB84shQoBVNHABOAOgDmnAEpw0EaibBwxS+GvVmwJuAxdTZDT1MdTkyAF8AbhA5EAhpGAl0ZFA6ExMIAHceHyg0ZBA6ADcICRYYwoxkbDp+Y1iAIxM6MABrTgBlaRapC2QYE2o4WJYIMGra+pApYxNuZotKOgm64ZAAKzQVACFmts66SjgAGSk4FamHGECAJgu17pNZgoa6BoFoCukTKRgAdTKMAAFsgABwABli3wgxj+zWkBW+dlMRXOsR8AEdqBIfFwFkt7rFjJQJP1Bms0L1+HAAIrUCDwQkgegNAEsYHIG6xAZ0CT8XoAYQglAJKCg0DRIGoxgAKm98kgaqtwuEgA).

```js
// In (expected):
const foo = {
  mutateAndGetPayload: (input, context) => (
    User.getResource(context).create(input)
  ),
};

// Out:
const foo = {
  mutateAndGetPayload: (input, context) =>
    User.getResource(context).create(input)
};
```
